[Merged by Bors] - Add FromReflect trait to convert dynamic types to concrete types

Davier opened the pull request:
Dynamic types (`DynamicStruct`, `DynamicTupleStruct`, `DynamicTuple`, `DynamicList` and `DynamicMap`) are used when deserializing scenes, but currently they can only be applied to existing concrete types. This leads to issues when trying to spawn non trivial deserialized scene.
For components, the issue is avoided by requiring that reflected components implement ~~`FromResources`~~ `FromWorld` (or `Default`). When spawning, a new concrete type is created that way, and the dynamic type is applied to it. Unfortunately, some components don't have any valid implementation of these traits.
In addition, any `Vec` or `HashMap` inside a component will panic when a dynamic type is pushed into it (for instance, `Text` panics when adding a text section).

To solve this issue, this PR adds the `FromReflect` trait that creates a concrete type from a dynamic type that represent it, derives the trait alongside the `Reflect` trait, drops the ~~`FromResources`~~ `FromWorld` requirement on reflected components, ~~and enables reflection for UI and Text bundles~~. It also adds the requirement that fields ignored with `#[reflect(ignore)]` implement `Default`, since we need to initialize them somehow.

cart commented:
This is definitely worth considering, but its also a pretty big change to the "scene/entity" lifecycle so I'll need to spend some time considering the design. I'll get to this as soon as I can, but I probably won't prioritize this for the 0.5 release.

NathanSWard commented:
To help track here are some related:

Issues:
#2215

PRs:
#2216 (attempts to fix)
#2359 (blocked)
#1429 (blocked)

Davier commented:
I updated the PR and trimmed it down, I hope it feels more like a small improvement to fix an issue and less like a big change to the "scene/entity" lifecycle :-)

## What I kept 
- add the `FromReflect` trait, which calls itself recursively on fields to construct a concrete copy of a possibly dynamic value
- derive `FromReflect` along with `Reflect`, add a `Default` bound to ignored fields
- manually implement `FromReflect` on `SmallVec`, `Vec`, `Hashmap` and `Cow<'static, str>`, fix `List::push()` on `SmallVec` and `Vec`
- add a test to deserialize containers which items do not implement `FromWorld`/`Default`
- change `FromWorld` bound to `FromReflect` in `ReflectComponent` (this could also be trimmed)

## What I removed
- reflect and enable deserialization on `Text` (that was my original motivation, but I will make a separate PR)
- use `FromReflect` in `Reflect::clone_value()` implementations  (I think it should be done, but we can talk about it later)

bjorn3 commented on crates/bevy_reflect/src/impls/std.rs:
```diff
@@ -1,4 +1,4 @@
-use crate as bevy_reflect;
+use crate::{self as bevy_reflect, FromReflect};
```
Can you please put the `FromReflect` in the use item below this use?

Davier replied:
Done!

Davier commented:
This PR was discussed here: https://discord.com/channels/691052431525675048/745805740274614303/869276523503435817

I think the main point was: "Reflect should be derivable for every type, including those which cannot be constructed from a combination of FromReflect and Default impls" (from cart)

Suggested changes are:
- don't use FromReflect in ReflectComponent
- split FromReflect into its own derive

For future reference, I wanted to use FromReflect in ReflectComponent because I had several cases where a Component could not correctly implement Default/FromWorld. For instance, components that contain an Entity must implement Default with a fake Entity in order to be reflected ([see Parent](https://github.com/bevyengine/bevy/blob/cf221f9659127427c99d621b76c8085c4860e2ef/crates/bevy_transform/src/components/parent.rs#L13)).

alice-i-cecile commented:
Have you seen the technique used in #2491? Perhaps we could use it here as well, to get elegant handling of the various initialization methods.

Davier commented:
> Have you seen the technique used in #2491? Perhaps we could use it here as well, to get elegant handling of the various initialization methods.

I don't think so unfortunately, there is no wrapping type like `Local` in this case. 

After thinking a bit more about it, introducing `FromReflect` (especially as a separate derive) may not be worth the effort if the only remaining use is in `List::push()`. We might as well "fix" the issues by requiring that types inside containers that need reflection must implement `Default`, the same way we already require that reflected components must implement `FromWorld`. It's a small change that could be reverted later if we find a better solution, but we would need to impl `Default` for `Entity` because of `Children`, which I don't like.

Davier commented:
Here is how the alternative looks like: https://github.com/Davier/bevy/commit/e621fcc41a597e53c2dcb04ab284f0542c32c58f
Please tell me which way you prefer, I'd be fine with either at this point.

alice-i-cecile commented:
> but we would need to impl Default for Entity because of Children, which I don't like.

Other users structs containing `Entity` are very common too. The other option here is actually to use niches to make `Option<Entity>` the same size as `Entity`. IIRC @cart and @BoxyUwU were discussing this in some other thread; a `NonZero` u64 would work fine for us here IMO and `Option<Entity>` comes up in enough other places that I want this anyways.

Then, we could change `Children` to store `Option<Entity>`, giving us the correct semantics: it starts explicitly uninitialized. 


Last time this came up, there was some frustration that we can't assign `None` to `u64::MAX` and the risk of off-by-one errors that can create. But given that the alternative is to create a dummy `Entity` id that is only used for defaulting or bloating the storage cost of every entity-containing component we need to serialize, that very much seems like the lesser of the evils.

Davier commented:
I hate having to handle the invalid `None` case everywhere even more than initializing to an invalid entity id TBH

NathanSWard commented on crates/bevy_reflect/src/impls/std.rs:
```diff
@@ -114,15 +144,29 @@ unsafe impl<T: Reflect> Reflect for Vec<T> {
+impl<T: FromReflect> FromReflect for Vec<T> {
+    fn from_reflect(reflect: &dyn Reflect) -> Option<Self> {
+        if let ReflectRef::List(ref_list) = reflect.reflect_ref() {
+            let mut new_list = Self::new();
```
```suggestion
            let mut new_list = Self::with_capacity(ref_list.len());
```

NathanSWard commented on crates/bevy_reflect/src/impls/smallvec.rs:
```diff
@@ -95,3 +97,20 @@ where
+{
+    fn from_reflect(reflect: &dyn Reflect) -> Option<Self> {
+        if let ReflectRef::List(ref_list) = reflect.reflect_ref() {
+            let mut new_list = Self::new();
```
```suggestion
            let mut new_list = Self::with_capacity(ref_list.len());
```

NathanSWard commented on crates/bevy_reflect/src/impls/std.rs:
```diff
@@ -230,6 +274,22 @@ where
+impl<K: FromReflect + Eq + Hash, V: FromReflect> FromReflect for HashMap<K, V> {
+    fn from_reflect(reflect: &dyn Reflect) -> Option<Self> {
+        if let ReflectRef::Map(ref_map) = reflect.reflect_ref() {
+            let mut new_map = Self::default();
```
```suggestion
            let mut new_map = Self::with_capacity(ref_map.len());
```

Davier commented:
I reverted the modifications to the reflection example, since `FromReflect` will not be expected to always be derived along `Reflect`. Should I make a specific example deriving it for a struct that is inside a container, or is that too niche?

Also, should I do the refactoring to deduplicate the code between `derive_reflect` and `derive_from_reflect`? Or is it be better left for another PR, like #2377?

cart reviewed (commented):
Ok so I like FromReflect / it seems like a relatively uncontroversial addition to the reflection toolset. I'm also convinced that using FromReflect in collections is a good solution in the short term. Just one small nit and I think we should merge this!

Davier commented on crates/bevy_reflect/bevy_reflect_derive/src/lib.rs:
```diff
@@ -740,3 +741,118 @@ pub fn external_type_uuid(tokens: proc_macro::TokenStream) -> proc_macro::TokenS
+        };
+
+        if let Some(ident) = meta_list.path.get_ident() {
+            if ident == REFLECT_ATTRIBUTE_NAME {
```
Removed in bb37b33

cart commented:
bors r+
